operators/pkg/instance-controller/container_logic.go

```diff
@@ -0,0 +1,206 @@
+				Value: strconv.FormatBool(examMode),
+			}, {
+				Name:  "HTTP_PORT",
+				Value: fmt.Sprintf("%d", httpPort),
+			}, {
+				Name:  "WEBSOCKIFY_PORT",
+				Value: fmt.Sprintf("%d", websockifyPort),
+			}},
+			SecurityContext: &contSecCtx,
+			LivenessProbe:   &noVncProbe,
+			// ReadinessProbe:  &noVncProbe,
+		},
+		{
+			Name:      "websockify",
+			Image:     r.DockerGuiWskfyImg + ":" + r.SvcContainersTag,
+			Resources: buildResRequirements(0.02, 50, resource.MustParse("50Mi")), // ~2MiB
+			Env: []v1.EnvVar{{
+				Name:  "WS_PORT",
+				Value: fmt.Sprintf("%d", websockifyPort),
+			}},
+			LivenessProbe: &websockifyProbe,
+			// ReadinessProbe: &websockifyProbe,
+		},
+		{
+			Name:            "tigervnc",
+			Image:           r.DockerGuiVncImg + ":" + r.SvcContainersTag,
+			Resources:       buildResRequirements(0.5, 50, resource.MustParse("500Mi")), // Mem depends on screen resolution, should easily manage up to 2k virtual screens
+			SecurityContext: &contSecCtx,
+			LivenessProbe:   &tigerVncProbe,
+			// ReadinessProbe:  &tigerVncProbe,
+		},
+		{
+			Name:  name,
+			Image: environment.Image,
+			Resources: buildResRequirements(
+				float32(environment.Resources.CPU),
+				environment.Resources.ReservedCPUPercentage,
+				environment.Resources.Memory,
+			),
+			SecurityContext: &contSecCtx,
+		},
+	}
+
+	template := &instance.Spec.Template
+	labels := map[string]string{
+		"name":                         name,
+		"crownlabs.polito.it/template": template.Namespace + "_" + template.Name,
+	}
+
+	return appsv1.DeploymentSpec{
+		Replicas: pointer.Int32Ptr(1),
+		Selector: &metav1.LabelSelector{
+			MatchLabels: labels,
+		},
+		Template: v1.PodTemplateSpec{
+
+			ObjectMeta: metav1.ObjectMeta{
+				Labels: labels,
+			},
+			Spec: v1.PodSpec{
+				Containers:      containers,
+				SecurityContext: &podSecCtx,
+			},
+		},
+	}
+}
+
+// CreateContainerEnvironment implements the logic to create all the different
+// Kubernetes resources required to start a containerized CrownLabs environment.
+func (r *InstanceReconciler) CreateContainerEnvironment(
+	instance *crownlabsv1alpha2.Instance,
+	environment *crownlabsv1alpha2.Environment,
+	namespace string,
+	name string,
+	vmStart time.Time) error {
+	ctx := context.TODO()
+
+	service, ingress, err := CreateInstanceExpositionEnvironment(ctx, r, instance, name, namespace)
+	if err != nil {
+		return err
+	}
+
+	depl := appsv1.Deployment{
+		ObjectMeta: BuildInstanceObjectMeta(name, namespace, environment, instance.Name),
+	}
+
+	if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &depl, func() error {
+		depl.Spec = buildContainerInstanceDeploymentSpec(name, instance, environment, r, 6080)
+		return ctrl.SetControllerReference(instance, &depl, r.Scheme)
+	}); err != nil {
+		r.setInstanceStatus(ctx, "Could not create deployment "+depl.Name+" in namespace "+depl.Namespace+": "+err.Error(), "Error", "GUIContainerDeplNotCreated", instance, "", "")
+		klog.Info(err, "")
```

Comment:
Why klog.Info on error? You should also log something besides the error itself

Reply:
It's actually something I initially used to find out why stuff wasn't working. It shouldn't be required anymore

Reply:
Perfect